### PR TITLE
Feature: variants and screen support

### PR DESCRIPTION
**Pull request by mhsdesign**
fixes: https://github.com/remcohaszing/monaco-tailwindcss/issues/7

image
image
image
--------

this feature adds a new console.warn:

image

probably because this variant:

image
-------
really tried to fix all ts types but no luck ...

### CODE REVIEW

**Comment by mhsdesign** [comment]
i think we should extend `{ State } from 'tailwindcss-language-service';` and make it jit only (so those props are non nullable)

**Comment by mhsdesign** [comment]
i fixed a lot of types by extending the State with a JitState (should be probably placed some place else) - but im unsure about https://github.com/remcohaszing/monaco-tailwindcss/pull/14/commits/75505b7b644bef2ea0f0332f732e91ddf9247876

**Comment by remcohaszing** [comment]
An alternative to defining our own `JitState` interface is to use non-null assertions.

```ts
state.modules!.postcssSelectorParser!.module!.className();
```

Neither solution is that great though, but unfortunately we have to deal with quirky upstream types.

**Comment by mhsdesign** [comment]
hmm i assumed this fancy ! would be doing that ^^ dindt now the name yet xD

but postcssSelectorParser fx is any 😶 because upstream likes any ^^

can we make `State` an union with different sub sets like Jit and non Jit and then ducktype?

```ts


interface NonJitState extends State {
   jit: false,
   jitContext: null,
}

interface JitState extends State {
  jit: true,
  jitContext: JitContext
}

interface BetterState NonJitState | JitState


function isJitState(state: State): state is JitState => state.jit

```

**Comment by mhsdesign** [comment]
@remcohaszing im kinda new to typescript (im only used to php types and psalm ^^)
can you please help me out with fixing this 😄 im eager to learn though ^^

**Comment by remcohaszing** [comment]
> @remcohaszing im kinda new to typescript (im only used to php types and psalm ^^) can you please help me out with fixing this smile im eager to learn though ^^

Ok that’s good to know! Then it’s also good to know you’ve gotten yourself in a pretty complex situation as far as types go here, because the upstream types simply aren’t that great.

The ideal solution is to provide fixes upstream, but for now we’re still in the process of finding out what those types are supposed to be. I don’t really have a good answer for this yet. Your solution to use the custom interface `JitState` isn’t ideal, but it’s definitely not bad either! Let’s just go with that for now.

I recommend you install a prettier and ESLint plugin for your editor and configure it to autofix on save. That should save some noise in the GitHub actions reports on the diffs.

I see there are only 2 TypeScript errors left in this PR. Just let me know if you need me to look into that.

File: src/getVariants.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  function escape(className: string): string {
+    const node = state.modules.postcssSelectorParser.module.className({ value: className });
+    return node.value;
+  }
+
+  const result: Record<string, string | null> = {};
+  // [name, [sort, fn]]
+  // [name, [[sort, fn]]]
+  for (const [variantName, variantFnOrFns] of state.jitContext.variantMap) {
+    const fns = (Array.isArray(variantFnOrFns[0]) ? variantFnOrFns : [variantFnOrFns]).map(
```
Comment (mhsdesign) [inline src/getVariants.ts:19]: i could not get this to work with type inference ...

**Comment by mhsdesign** [comment]
im now at the end of my wisdom ^^

File: src/getVariants.ts (hunk 1 above)
Comment (remcohaszing) [reply src/getVariants.ts:19]: I’ll have a look :+1:

Comment (mhsdesign) [reply src/getVariants.ts]: even their inline comment was wrong - the data dindt match 😂

**Comment by mhsdesign** [comment]
so i now could infer some further types by placing console.logs everywhere

**Comment by mhsdesign** [comment]
its green holy ....

**Comment by mhsdesign** [comment]
i dont know how upstream can live with that kindof bad types / no types at all ...

**Pull request merged**